### PR TITLE
Improve sts header parsing

### LINKED ISSUE #2810: Improve parsing of iam_request_headers (AWS IAM authentication)
I'm using the AWS JavaScript SDK to generate the required STS headers for the AWS IAM authentication login request. It generates something like this:

```
{
  "User-Agent": "aws-sdk-nodejs/2.28.0 darwin/v6.10.0",
  "Content-Type": "application/x-www-form-urlencoded; charset=utf-8",
  "X-Vault-AWS-IAM-Server-ID": "Bla",
  "X-Amz-Content-Sha256": "ABCDE",
  "Content-Length": 43,
  "Host": "sts.amazonaws.com",
  "X-Amz-Date": "20170605T191548Z",
  "x-amz-security-token": "very long token here"
}
```

When sending this (Base64 encoded, etc.) to Vault login, I'm getting the following error:

```
 "failed to JSON decode iam_request_headers \"{\\\"User-Agent\\\":\\\"aws-sdk-nodejs/2.23.0 linux/v6.10.2 exec-env/AWS_Lambda_nodejs6.10\\\",\\\"Content-Type\\\":\\\"application/x-www-form-urlencoded; charset=utf-8\\\",\\\"X-Vault-AWS-IAM-Server-ID\\\":\\\"bla\\\",\\\"X-Amz-Content-Sha256\\\":\\\"1234\\\",\\\"Content-Length\\\":43,\\\..."}\": json: cannot unmarshal number into Go value of type []string"
```

Since `Content-Length` is the only number, I removed it for testing. However, then I got: `json: cannot unmarshal string into Go value of type []string`

I looked into the [Vault source code](https://github.com/hashicorp/vault/blob/v0.7.2/builtin/credential/aws/path_login.go#L1103) and saw that the string is unmarshalled into `http.Header`, which defines an array of strings for each header value. 

This means I can't directly use the STS request header values, instead I have to convert any numbers to strings and any string objects to string arrays. This is a bit tedious, also the above error messages aren't straight forward, so I think this could be improved from a usability point of view. I also just wanted to post this here, so that people encountering the same error message can search here and find some information.

### PR DESCRIPTION
**Pull request by joelthompson**
I'm not sure if this is the best way to do so in go, so please let me know if there's a better way of accomplishing this!

Fixes #2810

### CODE REVIEW

File: builtin/credential/aws/path_login.go [hunk 1]
```diff
@@ -1407,7 +1406,38 @@ func parseGetCallerIdentityResponse(response string) (GetCallerIdentityResponse,
 	return result, err
 }
 
-func submitCallerIdentityRequest(method, endpoint string, parsedUrl *url.URL, body string, headers http.Header) (*GetCallerIdentityResult, error) {
+func parseIamRequestHeaders(headersB64 string) (*http.Header, error) {
```
Comment (jefferai) [inline builtin/credential/aws/path_login.go]: `http.Header` is a typedef for `map[string][]string` and as such it's a reference type -- return it directly, not as a pointer.

File: builtin/credential/aws/path_login.go [hunk 2]
```diff
@@ -1407,7 +1406,38 @@ func parseGetCallerIdentityResponse(response string) (GetCallerIdentityResponse,
...
+			for _, individualVal := range typedValue {
+				switch possibleStrVal := individualVal.(type) {
+				case string:
+					headers.Add(k, possibleStrVal)
+				default:
+					return nil, fmt.Errorf("Header %q contains value %q that has type %s, not string", k, individualVal, reflect.TypeOf(individualVal))
+				}
+			}
+		default:
+			return nil, fmt.Errorf("Header %q value %q has type %s, not string or []interface", k, typedValue, reflect.TypeOf(v))
```
Comment (jefferai) [inline builtin/credential/aws/path_login.go]: s/Header/header -- same above

**Comment by joelthompson** [comment]
Done.

> Unfortunately fixing this will require backing out half the changes you made in the first place...sorry!

No, thank you, I learned something new about go so it's great! :)

**Comment by jefferai** [comment]
Looks good -- thanks!

**Pull request merged**